### DISCUSSION TITLE
U8G2 HW SPI breaks after update from 2.6.2 to 3.6.0

### DISCUSSION THREAD

**Discussion by mattthegamer463** [earlephilhower/arduino-pico#1733] [Q&A] [answered]
Hi,

I've been using Arduino Pico 2.6.2 for a long time but recently updated it to current (3.6.0) because I saw there are some new features , but now my SPI LCD won't work.  My Hello World code is below:

```
#include <Arduino.h>
#include <U8g2lib.h>

#include <SPI.h>

//doesn't work
U8G2_ST7586S_ERC240160_F_4W_HW_SPI u8g2(U8G2_R0, /* cs=*/ 17, /* dc=*/ 20, /* reset=*/ 21);

//works
//U8G2_ST7586S_ERC240160_F_4W_SW_SPI u8g2(U8G2_R0, /*clock*/ 18,/*data*/ 19, /* cs=*/ 17, /* dc=*/ 20, /* reset=*/ 21);

void setup(void) {
 
  u8g2.begin();
  
}
void loop(void) {
  u8g2.clearBuffer();					// clear the internal memory
  u8g2.setFont(u8g2_font_logisoso16_tf);
  u8g2.drawStr(5,20,"Hello World!");	// write something to the internal memory
  u8g2.sendBuffer();					// transfer internal memory to the display
  delay(1000);  
} 
```

Declaring as software SPI works, but the refresh rate is very low.  Declaring HW SPI like I used to and the screen has no response.  If I compile using Arduino Mbed, HW SPI works, but my other projects use RPi_Pico_TimerInterrupt which Mbed errors on, saying it's not supported.

I'm using the default SPI0 and default pins, and it always worked on 2.6.2 so I'm not sure what's happened. Hopefully someone has an insight.

Thanks,
- Matt

**Comment by earlephilhower** [comment]
Sorry, but you're going to have to provide more info to get help.  What exactly is happening?  Is it hanging?  Report device not found?  Any other debug info?  (Turn on Tools->Debug Port->Serial and then Tools->Debug Level->SPI to capture).

You can also try bisecting things by moving somewhere between 2.6.2 and 3.6.0 (i.e. does 3.0.0 work but 3.1.0 fail) which would narrow things down tremendously.

**Reply by mattthegamer463** [reply]
Hi Earl, thanks for the reply.  It compiles, uploads and runs, the device responds to button presses, etc.  Just the display isn't initializing or showing any characters.  There are no other SPI devices in the device (unless the flash on the pico counts) and I almost exclusively use I2C devices normally so I don't have anything on hand to try (this is also a custom PCB).

If I configure U8G2 to use software SPI the screen works, if I enable U8G2 to use hardware SPI,  it doesn't work.  Both worked fine when on 2.6.2.

I also tried configuring the pins explicitly per here, no success: https://arduino-pico.readthedocs.io/en/latest/spi.html#

I haven't ever tried debugging SPI to capture, I will try that and also try going through the older versions to find when it breaks.  I'll also try probing the lines to look for data.

I've been using your core for over a year and I really appreciate what you've done for the Pico community.

**Reply by mattthegamer463** [reply]
Hi Earl,

I tried the SPI debug, this was all I got: 
```

SPI::transfer(00), cpol=1, cpha=1
SPI: read back 00
SPI::transfer(00), cpol=1, cpha=1
SPI: read back 00
SPI::transfer(00), cpol=1, cpha=1
SPI: read back 00
SPI::transfer(00), cpol=1, cpha=1
SPI: read back 00
SPI::transfer(00), cpol=1, cpha=1
SPI: read back 00
SPI::transfer(00), cpol=1, cpha=1
SPI: read back 00
SPI::transfer(00), cpol=1, cpha=1
SPI: read back 00
```
Not sure what to make of that.  This LCD doesn't have a MISO pin so there's no way for it to talk back, if that was what it was relying on.  I get the same results when it's working, on the lower version.

In version 3.2.2, it works, but in 3.3.0 and up it stops.  I'd keep using 3.2.2 but it doesn't appear to have the rp2040.rebootToBootloader() function that I want to use.

Thanks,
Matt

**Reply by earlephilhower** [reply]
Looking at https://github.com/earlephilhower/arduino-pico/compare/3.2.2...3.3.0 shows *no* SPI-related changes whatsoever in the core.

There was a Pico SDK 1.5.0->1.5.1 update: https://github.com/raspberrypi/pico-sdk/compare/1.5.0...1.5.1 which has a very minor SPI change in the RPI folks' code that seems to be innocuous https://github.com/raspberrypi/pico-sdk/commit/0121007c853a36cebcf78e591437d6c79c964d4b

**Reply by mattthegamer463** [reply]
I feel like I can rule out the U8G2 display library as it works with the earlier versions, and the hardware for the same reason.  I tried Googling for U8G2 issues with Arduino-core but didn't see anything relevant.  

Have you got any other ideas for things I can try, to try to get more information?  I have a RP2040 datalogger I built with SPI SD card, would that be worth experimenting with?

**Reply by earlephilhower** [reply]
If you have a logic analyzer, can you check the outputs from 3.2.2 and 3.3.0 and see where the difference is?  It all boils down to what's on those wires.

Release 3.3.0 came out 4 months ago, and I'm pretty sure people have used the SPI ports in that time.

**Reply by mattthegamer463** [reply]
I can do that.  I'll post what I find.  Thanks.

**Reply by earlephilhower** [reply]
Awesome, that should really help.  Also, I just realized that I've personally used `SPI` a lot with the wired Ethernet PR (all SPI, all the time!) for the latest release!  :laughing:

**Reply by mattthegamer463** [reply]
Hi Earl,

From what I can see, the working and non working screen draw blocks match, but the initialization code is different.  The blocks in beige show the init stuff is the same in 3.3.0 SW SPI and 3.2.2 HW SPI.

<img width="1715" alt="3 3 0 vs 3 2 2 compare" src="https://github.com/earlephilhower/arduino-pico/assets/45496113/e197f37c-877b-47a2-9b57-520a22cee9fe">

I'm not sure why it would be different but this must explain the issue.  I guess it's got something to do with how U8G2 interacts with 3.3.0+ but I haven't the foggiest about why it would change.  I may need to post to the U8G2 page to see if the creator of U8G2 has any insight.

**Reply by earlephilhower** [reply]
Very interesting!

In hex, (good) `0x01_11_28_c0 >> 1`  == (bad) `0x00_88_94_60`.  Your trace shows a 1-bit shift in the recorded SPI for 3.3.0 vs the others.

Can you get a good capture of the first two bytes on 3.3.0 so we can see the actual clock and bits?

Also, w/the SPI debug enabled, can you grab the startup output (i.e. from the beginning of time until the first few SPI bytes are written?  That would show the SPI configuration and if the library is doing weird things like reconfiguring the SPI timings while running.

And finally, what is the SPI HW frequency here?  Could it just be a setup/hold issue where it was marginal to begin with but is now pushed over the edge?

**Reply by earlephilhower** [reply]
Also, check for clock glitching from reset until the first SPI byte.  Could be something as simple as the reset state of a pin toggled while coming up, causing the SPI receiver to think the 1st bit was sent at T=0.  Then when the real SPI data comes along everything looks shifted 1-bit by the receiver.

FWIW, I've run SPI @ 30MHZ for the wired Ethernet adapters w/o incident as part of testing the last release.  https://github.com/earlephilhower/arduino-pico/pull/1703#issuecomment-1718621708

**Reply by mattthegamer463** [reply]
Hi Earl,

Here's the very start of initialization on 3.3.0
<img width="1661" alt="3 3 0 startup" src="https://github.com/earlephilhower/arduino-pico/assets/45496113/75c348ce-8e43-4733-b111-73849cf71b0c">

Now, I was having a hard time getting the serial monitor to capture the start of initialization because the port is occupied.  So I tried moving the display begin() function call to the loop instead of setup, so it would repeat it.  This made the display start working.  I tried adding a delay before the begin() in setup, to see if that helped. It didn't.  What does make it work however, is running the begin() function twice.

<img width="1641" alt="3 3 0 double init" src="https://github.com/earlephilhower/arduino-pico/assets/45496113/65f4b2e7-39d4-42fe-a33a-9eb176945fb6">

I've updated to 3.6.0 and it's working as long as I do the initialization twice.

I did manage to get the initialization:

```
SPI::begin(0), rx=16, cs=17, sck=18, tx=19
SPI::beginTransaction(clk=4000000, bo=MSB)
SPI: initting SPI
SPI: actual baudrate=4000000
SPI: IRQ masks before = 00000000 00000000 00000000 00000000
SPI: IRQ masks after = 00000000 00000000 00000000 00000000
SPI::endTransaction()
SPI: IRQ masks = 00000000 00000000 00000000 00000000
SPI::beginTransaction(clk=8000000, bo=MSB)
SPI: deinitting currently active SPI
SPI: initting SPI
SPI: actual baudrate=8000000
SPI: IRQ masks before = 00000000 00000000 00000000 00000000
SPI: IRQ masks after = 00000000 00000000 00000000 00000000
SPI::endTransaction()
SPI: IRQ masks = 00000000 00000000 00000000 00000000
SPI::beginTransaction(clk=8000000, bo=MSB)
SPI: Reusing existing initted SPI
SPI: IRQ masks before = 00000000 00000000 00000000 00000000
SPI: IRQ masks after = 00000000 00000000 00000000 00000000
SPI::transfer(01), cpol=1, cpha=1
SPI: read back 00
SPI::transfer(11), cpol=1, cpha=1
SPI: read back 00
```

The SPI Frequency is default, the clock speed is 133MHz.  I tried downclocking to 50MHz and the logic analyzer showed a change, but it didn't fix it.

It's interesting if there is a 1 bit shift causing all the data sent during init to be wrong, but when it's run a 2nd time it doesn't have that problem.

**Reply by earlephilhower** [reply]
Looking at your close-up first image, you can clearly see the clock-stretched first (invalid?) bit.  I wonder if there's a race condition somewhere in the library or SPI code in the SDK 1.5.1 vs 1.5.0 WRT the device enable/disable stuff they added.

The SDK codepath is called 2x because it starts up at default 4MHZ and then is changed to 8MHZ by the GFX library.  The core `SPI` class is smart enough not to reset the SPI device if all settings are the same between transactions, so we can try setting the defaults here to 8MHZ and see if avoiding that clock speed change avoids the extra clock transition.

Edit
https://github.com/earlephilhower/arduino-pico/blob/41b0686aec4b0655e518e3ef98379131ad3773fa/libraries/SPI/src/SPI.cpp#L322-L324
to
````
_spis = SPISettings(8000000, MSBFIRST, SPI_MODE0);
beginTransaction(_spis);
endTransaction();
````

**Reply by mattthegamer463** [reply]
Hi Earl,

I think I implemented the change correctly, the file was in C:\Users\mattt\AppData\Local\Arduino15\packages\rp2040\hardware\rp2040\3.6.0\libraries\SPI\src

Logic analyzer show it's still running at the same clock speed but there is still the bit shift during init.

EDIT: 

Initializing twice no longer works either, LA shows the 1 bit shift is occuring for both init calls now.

EDIT 2: 

If I undo the spi.cpp change you recommended, the 2nd init starts to work again.

**Reply by earlephilhower** [reply, marked as answer]
Gotta say that I'm pretty stumped now.  No issues with other SPI users so there still must be something specific about this use case.  I'm thinking it might not be the protocol so much as something else here.

SPI interfaces normally have a CS, yes?  Are you sure the CS pin is disabled until the GFX lib starts communicating?  There might always be some glitchiness as the SPI unit comes online and takes over the pins, but it would be ignored since CS would be deasserted until a real command comes.  Can you capture the CS pin as well and see that it's disabled until the 1st valid bit of data?

**Reply by mattthegamer463** [reply]
Hi Earl,

In the LA captures, line D0 is the CS pin for the device.  It's GP17 on the Pi which is "SPI0 CSn" and I don't touch it, the SPI and LCD lib are doing whatever they wish with it when I call the begin() function.  It doesn't look like it's doing anything weird as far as I can see from the captures.

I greatly appreciate your help with this, since it is working by calling the initialization twice, I'm happy to call this fixed and chalk up the root cause to the gremlins.
